identification_suggester.py

```diff
@@ -0,0 +1,244 @@
+         
+<<<<<<< HEAD
+<<<<<<< HEAD
+        program = self.frontdoor_program()
+=======
+        program = self._frontdoor_program()
+>>>>>>> identification
+=======
+        program = self._frontdoor_program()
+>>>>>>> validator
+
+        frontdoor_set : Set[str] = []
+
+        for confounder in confounders: 
+            if confounder != treatment and confounder != outcome:
+                success = False
+                while not success: 
+                    try:
+                        output = program(treatment=treatment, outcome=outcome, confounder=confounder, llm=llm)
+
+                        z = re.findall(r'<answer>(.*?)</answer>', output['frontdoor_set'], re.DOTALL)
+
+                        if len(z) > 0: 
+                            if z[0] == 'A' or z[0] == 'A. Yes':
+                                frontdoor_set.append(confounder)
+                            
+                            success = True
+
+                        else:
+                            success = False
+
+                    except KeyError:
+                        success = False
+                        continue 
+
+        return frontdoor_set
+    
+    def suggest_iv(self, confounders: List[str], treatment: str, outcome: str, llm: guidance.llm):
+         
+<<<<<<< HEAD
+<<<<<<< HEAD
+        program = self.iv_program()
+=======
+        program = self._iv_program()
+>>>>>>> identification
+=======
+        program = self._iv_program()
+>>>>>>> validator
+        
+        iv_set : set[str] = []
+
+        for confounder in confounders: 
+            if confounder != treatment and confounder != outcome:
+                success = False
+                while not success: 
+                    try:
+                        output = program(treatment=treatment, outcome=outcome, confounder=confounder, llm=llm)
+
+                        z = re.findall(r'<answer>(.*?)</answer>', output['iv_set'], re.DOTALL)
+
+                        if z: 
+                            if z[0] == 'A' or z[0] == 'A. Yes':
+                                iv_set.append(confounder)
+                            
+                            success = True
+
+                        else:
+                            success = False
+
+
+                    except KeyError:
+                        success = False
+                        continue 
+
+        return iv_set
+    
+<<<<<<< HEAD
+<<<<<<< HEAD
+    def backdoor_program(self):
+=======
+    def _backdoor_program(self):
+>>>>>>> identification
+=======
+    def _backdoor_program(self):
+>>>>>>> validator
+            
+        return guidance(    
+            '''
+            {{#system~}}
+            You are a helpful assistant on causal reasoning and arctic sea ice and atmosphere sciences. Your goal is to answer questions factually and concisely about whether and how a causal effect can be estimated using your domain knowledge on artic sea ice and atmosphere sciences.
```

Comment:
Given that this is a generic suggestor, let's remove the parts of the prompt that says the agent is an expert in arctic sea ice and atmospheric sciences, etc.

Reply:
Hmm... adding in an optional "what domain is this about" argument to the protocol that gets integrated into prompts might be useful.